src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -4543,8 +4586,14 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
             return;
         }
         this->visit_expr(*x.m_value);
-        ASRUtils::EXPR(tmp);
-        tmp = nullptr;
+
+        // If tmp is a statement and not an expression
+        // never cast into expression using ASRUtils::EXPR
+        // Just ignore and exit the function naturally.
+        if( !ASR::is_a<ASR::stmt_t>(*tmp) ) {
+            ASRUtils::EXPR(tmp);
+            tmp = nullptr;
+        }
```

Comment:
What is the purpose of these lines, both the original and the new ones?

Are we checking that the result is an expression? If so, you can use an assert for `ASR::is_a` (or perhaps `down_cast`, since `is_a` might require a particular class).

Reply:
So the scene is that `goto.end` is represented as `[(Expr (Attribute (Name goto Load) end Load))]` at AST level. So, now when we will complete visiting `Attribute`, `tmp` will either point to a ASR expression or a ASR statement (because `GoTo` in ASR is a statement). So in case we get something like, `goto.label` then `tmp` will point to a statement in which case we should exit the function naturally and the `transform_stmts` before this node will include that GoTo statement. However if `tmp` is not pointing to a statement then I replaced `ASRUtils::EXPR(tmp)` with an `LFORTRAN_ASSERT`.

Reply:
> If so, you can use an assert for `ASR::is_a` (or perhaps `down_cast`, since `is_a` might require a particular class).

Assert with `ASR::is_a` works fine.

Reply:
Can you please document this as a comment? Next person to see that code will have the same question, including me tomorrow.